demos/qml_optimization.ipynb

```diff
@@ -0,0 +1,469 @@
+        "\n",
+        "In this example, we'll show you how to use [JAX](https://jax.readthedocs.io), an autodifferentiable machine learning framework, and [Optax](https://optax.readthedocs.io/), a suite of JAX-compatible gradient-based optimizers, to optimize a PennyLane quantum machine learning model.\n",
+        "\n",
+        "## Set up your model, data, and cost\n",
+        "\n",
+        "Here, we will create a simple QML model for our optimization. In particular:\n",
+        "\n",
+        "- We will embed our data through a series of rotation gates.\n",
+        "- We will then have an ansatz of trainable rotation gates with parameters `weights`; it is these values we will train to minimize our cost function.\n",
+        "- We will train the QML model on `data`, a `(5, 4)` array, and optimize the model to match target predictions given by `target`."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 3,
+      "metadata": {
+        "id": "iKtDMXjENR7X"
+      },
+      "outputs": [],
+      "source": [
+        "import pennylane as qml\n",
+        "from jax import numpy as jnp\n",
+        "import jax\n",
+        "import optax\n",
+        "import catalyst"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 4,
+      "metadata": {
+        "id": "tiK-itpFNVwE"
+      },
+      "outputs": [],
+      "source": [
+        "n_wires = 5\n",
+        "data = jnp.sin(jnp.mgrid[-2:2:0.2].reshape(n_wires, -1)) ** 3\n",
+        "targets = jnp.array([-0.2, 0.4, 0.35, 0.2])\n",
+        "\n",
+        "dev = qml.device(\"lightning.qubit\", wires=n_wires)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 5,
+      "metadata": {
+        "id": "w4QPSXdysqJx"
+      },
+      "outputs": [],
+      "source": [
+        "@qml.qnode(dev)\n",
+        "def circuit(data, weights):\n",
+        "    \"\"\"Quantum circuit ansatz\"\"\"\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def data_embedding(i):\n",
+        "        qml.RY(data[i], wires=i)\n",
+        "\n",
+        "    data_embedding()\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def ansatz(i):\n",
+        "        qml.RX(weights[i, 0], wires=i)\n",
+        "        qml.RY(weights[i, 1], wires=i)\n",
+        "        qml.RX(weights[i, 2], wires=i)\n",
+        "        qml.CNOT(wires=[i, (i + 1) % n_wires])\n",
+        "\n",
+        "    ansatz()\n",
+        "\n",
+        "    # we use a sum of local Z's as an observable since a\n",
+        "    # local Z would only be affected by params on that qubit.\n",
+        "    return qml.expval(qml.sum(*[qml.PauliZ(i) for i in range(n_wires)]))\n",
+        "\n",
+        "circuit = qml.qjit(catalyst.vmap(circuit, in_axes=(1, None)), autograph=False)"
```

Comment:
Nice, putting all the features together 😄  I guess the only thing missing is autograph, is that because of specific issues right now, or because non-autograph is generally more robust?

Reply:
Unfortunately this is because of the issue that AutoGraph is converting the following if statement into a `catalyst.cond`:

https://github.com/PennyLaneAI/catalyst/blob/afbe7035ccb84cf5cbff2f0b1985e99349fb2e64/frontend/catalyst/pennylane_extensions.py#L2265-L2266

which is causing vmap to break (it tries to dispatch to JAX when it shouldn't).

This is the only thing stopping me from using Autograph!

Reply:
This example seems to work fine with autograph and vmap, there must be another element 🤔 
```py
@vmap
@qml.qnode(qml.device("lightning.qubit", wires=1))
def f(x):
    qml.RY(x, wires=0)
    return qml.probs()

@qjit(autograph=True)
def g():
    x = jnp.array([0., 1., 2., 3.])
    return f(x)

g()
```

Reply:
The problem appears to be the use of the nested `qjit` decorators.

While it is supported in the sense that inner qjit decorators are supposed to be transparent, it appears that in conjunction with certain options this causes issues (here autograph). If the support were working as intended though, the inner `autograph=True` flag would be fully ignored anyways.

Unfortunately, it appears natural for users to place `qjit` all over the place as I've seen the pattern before, whereas it really is it best thought of as the (unique!) entry point marker to the compiled program.

This could be an indication that the feature I suggested yesterday is required to better match user expectations @josh146

Reply:
> This example seems to work fine with autograph and vmap, there must be another element 🤔

Oh yes! I think @maliasadi also noticed this yesterday on slack. It only occurs if you directly stack `qjit(vmap(func), autograph=True)`. 

If you instead do `qjit(func, autograph=True)`, where `func` internally calls `vmap`, this works fine.

Ah, I guess what is happening here is that since `vmap` is being passed to `qjit`, autograph is directly applying to `vmap` as the entry point?

Reply:
In this particular case (for me as a user), since I was just trying to do `qjit(vmap(circuit))` to test if my circuit executed correctly, it seemed much more natural to me than having an additional level of function nesting 😢

Reply:
yes, that is a quick workaround for this issue by using `qjit` as the decorator instead. As being said, the root of this issue is about AG trying to convert all control flow operations to QJIT compatible ones. I also tried to use function wrappers around these if-conditions and trigger them as function calls (to see the AG behaviour in the case of nested func calls) but it didn't fix the issue.

Reply:
@maliasadi I was chatting to @dime10, and adding `vmap` to this list here might solve this issue:

https://github.com/PennyLaneAI/catalyst/blob/3606aec53842552919c19e65bbad7feb72daf278/frontend/catalyst/ag_primitives.py#L523-L531

Reply:
So unfortunately this situation is not covered by that hotfix. The big issue is the "primary" function we are feeding into qjit is an internal wrapper, and Catalyst internal functions should typically not be converted. On the other hand, the function fed into qjit must be converted by autograph since that is the starting. I think we can work around the issue with some small-ish amount of work, but for this PR we can close this comment chain I think.

Reply:
In that case removing the flag might make sense:
```suggestion
        "circuit = qml.qjit(catalyst.vmap(circuit, in_axes=(1, None)))"
```